retrieval/target.go

```diff
@@ -359,9 +364,15 @@ func (t *Target) scrape(sampleAppender storage.SampleAppender) (err error) {
 	}()
 
 	for samples := range t.ingestedSamples {
-		for _, s := range samples {
-			s.Metric.MergeFromLabelSet(baseLabels, clientmodel.ExporterLabelPrefix)
-			sampleAppender.Append(s)
+		if t.proxy {
+			for _, s := range samples {
+				sampleAppender.Append(s)
```

Comment:
I think there is a misunderstanding.

Here the three postulated modes explained with more words:

"Normal": Server will auto-attach `job` and `instance` labels. If any of those labels already exists, an `exporter_job` or `exporter_instance` label, respectively, is attached instead.

"Pushgateway": Same as normal, but if  a `job` or `instance` label already exists, no additional label is attached _even if one of those labels has an empty value_. The empty label will then be discarded (before storing the sample).

"Federation": No label is auto-attached.

"Federation" mode is needed because in federation, we do not want to change the metric in any way, even if it has e.g. no `instance` labels (as it is the case for metrics that have been exposed by the Pushgateway with an empty `instance` label).

As @brian-brazil said, we could make the server to attach empty `instance` labels to all labels without `instance` label if exposed via the `/federation` endpoint, similar to what the Pushgateway is doing. Then we also needed to make that configurable, as we did for the Pushgateway.

In case of the Pushgateway, we could not use "Federation" mode because we still want to attach `instance` labels to metrics of the Pushgateway itself. We do not have that problem for the `/federation` endpoint. "Federation" mode is easier to understand than "Pushgateway" mode. So one could argue we should have "Federation" mode to not make federation needlessly difficult to understand for users that never touch Pushgateway. On the other hand, by doing everything via "Pushgateway" mode, we need one less configuration option for the scrape config.